### ISSUE TITLE
S4049 reports interface functions or functions that require non-trivial calculation as functions that should be properties

### ISSUE THREAD

**Issue by StingyJack** [SonarSource/sonar-dotnet#4624]
Both of these interface functions are triggering S4049, This seems inappropriate to complain about an interface definition that knows nothing of the complexity required to implement. Even if it complained about the implementation (which Sonar does not), it should not be considering every function with a name starting with "Get" to be equivalent to a simple property (Complex properties are a code smell as properties imply minimal - if any - calculation). The data returned for these is calculated at runtime so a function makes more sense than a property.
``` csharp
    public interface IConnectionStringProvider
    {
        string GetSharedConfigDbConnString();
        
        List<string> GetSecondaryRegionConfigDbConnStrings();
     ...   
```

I expect this warning to be flagging things like the ones cited in the example (note this is a class not an interface)

``` csharp
namespace MyLibrary
{
    public class Foo
    {
        private string name;

        public string GetName()  // Noncompliant
        {
            return name;
        }
    }
}
```

**Comment by andrei-epure-sonarsource** [comment]
Hello @StingyJack and thanks for the issue. Indeed, at an interface level, one cannot tell the complexity of the implementation. This is also tied to #3200.

What puts me in difficulty is when the interface specifies a `GetFoo()` for which the implementation is `GetFoo() => foo;` - in this case we'd want to raise, however the rule currently doesn't raise when the method is an override or an interface implementation. In this case, we'd end up with a false negative.

Also, I would argue that for complex methods, `Get` wouldn't be the best verb - it should be `Compute` or a verb implying there's some complex (possibly expensive) logic behind the scenes, it's not just a "getter method" which should be just a property. 

WDYT?

**Comment by StingyJack** [comment]
I think that this is kind of a useless rule. There's nothing to be gained in quality or readability and it easily risks misleading callers as to the cost of using a property.  Its is also based on a limited definition of "Get" that assumes the functionality must be like a property getter, but there are so many other functions that go get something and that dont require a parameter. `GetCacheInstance()`, `GetClient()`, etc. 

Having a different setting for cyclomatic complexity that would be applied to properties would be a better rule. For example, properties with a CC of >3 should be considered for conversion to methods

**Comment by andrei-epure-sonarsource** [comment]
> I think that this is kind of a useless rule.

This is why it's not part of the default SonarWay Quality Profile - it's debatable. Some people find it useful, though and activate it in their quality profiles  (including ourselves).

> For example, properties with a CC of >3 should be considered for conversion to methods

That's a good suggestion, I opened  #4674 to keep track of it.

**Comment by csaba-sagi-sonarsource** [comment]
I will close this issue, as @andrei-epure-sonarsource already mentioned, some people find this rule useful and it is not in the default SonarWay Quality Profile.

And thanks to your suggestion there is #4674 to take Cognitive/Cyclomatic Complexity into consideration.

**Comment by StingyJack** [comment]
@csaba-sagi-sonarsource @andrei-epure-sonarsource  - I think you may have missed that this is happening on interface definitions, not the actual implementations. This is where its useless.
